com.sap.cloud.lm.sl.mta/src/main/java/com/sap/cloud/lm/sl/mta/builders/v1/ModuleChainBuilder.java

```diff
@@ -0,0 +1,95 @@
+package com.sap.cloud.lm.sl.mta.builders.v1;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.sap.cloud.lm.sl.mta.handlers.v1.DescriptorHandler;
+import com.sap.cloud.lm.sl.mta.handlers.v1.ModuleComparator;
+import com.sap.cloud.lm.sl.mta.model.v1.DeploymentDescriptor;
+import com.sap.cloud.lm.sl.mta.model.v1.Module;
+
+public class ModuleChainBuilder implements com.sap.cloud.lm.sl.mta.builders.ModuleChainBuilder {
+
+    protected DeploymentDescriptor descriptor;
+    protected DescriptorHandler handler;
+    protected String dependencyTypeProperty;
+    protected String hardDependencyType;
+
+    public ModuleChainBuilder(DeploymentDescriptor descriptor, DescriptorHandler handler,
+        String dependencyTypeParameter, String hardDependencyType) {
+        this.descriptor = descriptor;
+        this.handler = handler;
+        this.dependencyTypeProperty = dependencyTypeParameter;
+        this.hardDependencyType = hardDependencyType;
+    }
+
+    @Override
+    public List<? extends Module> build() {
+        Map<Module, Set<String>> modulesAndDeploymentDependencies = getModulesAndDeploymentDependencies(descriptor);
+        List<Entry<Module, Set<String>>> modulesAndDeploymentDependenciesSorted = modulesAndDeploymentDependencies.entrySet()
+            .stream()
+            .collect(Collectors.toList());
+
+        Collections.sort(modulesAndDeploymentDependenciesSorted, getModuleComparator(dependencyTypeProperty, hardDependencyType));
+
+        return modulesAndDeploymentDependenciesSorted.stream()
+            .map(Entry::getKey)
+            .collect(Collectors.toList());
+    }
+
+    protected Map<Module, Set<String>> getModulesAndDeploymentDependencies(DeploymentDescriptor descriptor) {
+        Map<Module, Set<String>> result = new LinkedHashMap<>();
+        for (Module module : descriptor.getModules1()) {
+            result.put(module, getModuleDependencies(descriptor, module));
+        }
+        return result;
+    }
+
+    protected Set<String> getModuleDependencies(DeploymentDescriptor descriptor, Module module) {
+        return getModuleDependencies(descriptor, module, new HashSet<String>());
+    }
+
+    protected Set<String> getModuleDependencies(DeploymentDescriptor descriptor, Module module, Set<String> seenModules) {
```

Comment:
Try to extract this in a separate class and see if it's better. seenKeys looks like a field for that class to me.